src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```
@@ -197,23 +253,71 @@ String getTopicNameFromNamespace(final String prefix, final BsonDocument namespa
+            ChangeStreamDocument<Document> firstResult = changeStreamCursor.tryNext();
+            if (firstResult != null) {
+                cachedResult = new BsonDocumentWrapper<>(firstResult, ChangeStreamDocument.createCodec(Document.class,
+                        MongoClientSettings.getDefaultCodecRegistry()));
+            }
+            cachedResumeAfter = firstResult != null ? firstResult.getResumeToken() : changeStreamCursor.getResumeToken();
+            changeStreamCursor.close();
+            return true;
+        }
+        return false;
+    }
+
     /**
      * Returns the next document to be delivered to Kafka.
      *
      * <p>
-     * Return the next result from the change stream cursor. Creating the cursor if necessary.
+     * <ol>
+     * <li>If copying data is in progress, returns the next result.</li>
+     * <li>If copying data and all data has been copied and there is a cached result return the cached result.</li>
+     * <li>Otherwise, return the next result from the change stream cursor. Creating a new cursor if necessary.</li>
+     * </ol>
+     *
      * </p>
      *
      * @return the next document
      */
     private Optional<BsonDocument> getNextDocument() {
+        if (isCopying.get()) {
+            Optional<BsonDocument> result = copyDataManager.poll();
+            if (result.isPresent() || copyDataManager.isCopying()) {
+                return result;
+            }
+
+            // No longer copying
+            isCopying.set(false);
+            if (cachedResult != null) {
+                result = Optional.of(cachedResult);
+                cachedResult = null;
+                return result;
+            }
+        }
+
         if (cursor == null) {
             cursor = createCursor(sourceConfig, mongoClient);
         }
 
         try {
             return Optional.ofNullable(cursor.tryNext());
         } catch (Exception e) {
+            cursor.close();
```

Don't we want to propagate this exception?

---

No, as the cursor will be recreated. One scenario the connector has to handle is if a collection being monitored doesn't yet exist or is deleted and recreated.